accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,215 @@
+
+```log
+~\.nuget\plugins\tools
+❯ .\dotnet-codeartifact-creds.exe
+Required command was not provided.
+
+Usage:
+  dotnet-codeartifact-creds [options] [command]
+
+Options:
+  --version         Show version information
+  -?, -h, --help    Show help and usage information
+
+Commands:
+  install      Installs the AWS CodeArtifact NuGet credential provider into the NuGet plugins folder.
+  configure    Sets or Unsets a configuration
+  uninstall    Uninstalls the AWS CodeArtifact NuGet credential provider from the NuGet plugins folder.
+```
+
+## Explanation
+
+### Functional explanation
+
+A deployment solution for .NET is the [.NET tools](https://learn.microsoft.com/dotnet/core/tools). These tools provide a seamless installation and management experience for NuGet packages in the .NET ecosystem. The use of .NET tools as a deployment mechanism has been a recurring request from users and internal partners who need to authenticate with private repositories. Currently, this solution works for the Windows .NET Framework. However, the goal is to extend this support cross-platform for all supported .NET runtimes.
+
+By implementing this specification, we offer plugin authors the option to use .NET tools for plugin deployment. They can install these as a `tool path` global tool. This eliminates the need to maintain separate versions for `.NET Framework` and `.NET Core`. It also simplifies the installation process by removing the necessity for plugin authors to create subcommands like `codeartifact-creds install/uninstall`.
+
+The ideal workflow for repositories that access private NuGet feeds, such as Azure DevOps, would be:
+
+1. Ensure that the dotnet CLI tools are installed.
+2. Execute the command `dotnet tool install Microsoft.CredentialProviders --tool-path "%UserProfile%/.nuget/plugins/tools"` on the Windows platform. For Linux and Mac, run `dotnet tool install Microsoft.CredentialProviders --tool-path $HOME/.nuget/plugins`.
+3. Run `dotnet restore --interactive` with a private endpoint. It should 'just work', meaning the credential providers installed in step 2 are used during credential acquisition and are used to authenticate against private endpoints.
+
+The setup instructions mentioned in step 2 are platform-specific, but they are simpler compared to the current instructions for installing credential providers.
+
+I proposed using a `tool path` .NET tool because, by default, .NET tools are considered [global](https://learn.microsoft.com/dotnet/core/tools/global-tools). This means they are installed in a default directory, such as `%UserProfile%/.dotnet/tools` on Windows, which is then added to the PATH environment variable.
+- Global tools can be invoked from any directory on the machine without specifying their location.
+- One version of a tool is used for all directories on the machine.
+However, NuGet cannot easily determine which tool is a NuGet plugin.
```

Comment:
Is this the crux of the issue and the cause for the "tool path" approach?

Reply:
In how many scenarios would there be more than one credential provider? The current approach seems predicated on a plugin approach that needs to discover some unknown number of tools with unknown names. I suspect that the VAST number of nuget operations would require zero or one credential provider tools, and a miniscule would require two or more.

Consider whether a combination of:
1. A means to specify the group of cred provider command names to try (e.g. the `<packageSourceCredentials>` proposal below)
1. A well-known default credential provider name to look for (e.g. "NuGetCredentialProvider") if no list is specified
would make for a simple, understandable approach.

Reply:
I haven't read the whole proposal yet, but lots of tools, including git, cargo, even dotnet I think use a convention-based approach for custom subcommands -- if there's a program named `git-foo-bar` on the `PATH`, then `git foo-bar --some-option` invokes `git-foo-bar --some-option` with some environment variables set up. Could something similar work here -- every program named `nuget-credential-provider-*` on the PATH is assumed to be a credential provider. For Windows, you'd iterate over PATH and take files with the right prefix and extensions as listed in PATHSPEC, and for Unixlikes, you'd iterate over PATH and take files with the right prefix and the execute bit set

Reply:
This could easily be extended to dotnet local tools as well, even though I don't think they end up on the PATH.

Reply:
If a credential provider is somehow specified by name, you wouldn't bother iterating, you'd just do whatever is necessary to see if it's a local tool, and if it's not, just invoke the program by name (so that the OS applies its usual search rules)

Reply:
> Consider whether a combination of:
> 
> 1. A means to specify the group of cred provider command names to try (e.g. the `<packageSourceCredentials>` proposal below)
> 2. A well-known default credential provider name to look for (e.g. "NuGetCredentialProvider") if no list is specified

Phil, as you mentioned in the comment, I have already added this as a future possibility. However, given the limited number of plugin implementations currently available, it would be hard to imagine the need for updating the `NuGet.Config` file schema to support plugin settings per repository. Please let me know if I am missing something.